### PR TITLE
Improved resampling checks

### LINKED ISSUE #10: Dealing with data frequencies close to the calendar frequency.
Currently we check if the data frequency is the same or higher than the calendar frequency it will be resampled to, and if not, warn the user. However, resampling data to a frequency close to the original frequency can cause issues.

For example; resampling data with an interval of 2 days to a 3 day calendar could cause issues such as aliasing.

A possible solution could be to use a Nyquist-like criterion: warn if the data sample rate is larger than half the calendar interval.

### PR DESCRIPTION
**Pull request by BSchilperoort**
In this PR:

- A robust check for frequency mismatch between the data and calendar is implemented (closes #10).
- Removed pytest.fixture(autouse=True), it was used incorrectly, and lead to over-testing. 
- Resampling is changed so if the user inputs a DataArray, they get a DataArray back.
  - This lead to a change in the definition of "bounds" of the resampling output. Instead of "interval" with dims (anchor_year, i_interval, bounds), there are now two coords (left_bound and right_bound) with dims (anchor_year, i_interval). This change was required to preserve the interval datetimes in the DataArray output.

Additionally, I made some small changes to the ruff configuration, the tests are checked now as well, but docstyles are ignored for that folder.

### CODE REVIEW

**Comment by BSchilperoort** [comment]
Hm. Turns out that there's a problem with the traintest notebook. I think we'll have to check the xarray/pandas input data for reserved coordinate/column names, e.g.:

anchor_year, i_interval, left_bound, right_bound, is_target (rename from "target").

An error should be raised if the data already contains one or more of these

**Review by geek-yang — CHANGES_REQUESTED** [review]
Nice work! The changes look nice and they help a lot to improve the "quality of life". Only got two small points:

- In `utils.py`, we have defined 
```py
PandasData = (pd.Series, pd.DataFrame)
XArrayData = (xr.DataArray, xr.Dataset)
```
But they are not used in some places. It would be nice to either use them when they could be applied, or just delete them and use the pandas and xarray object type.

- The checks for frequency works for most of the cases, but not for monthly calendar (see the minimum example below)

**Comment by geek-yang** [comment]
The checks for frequency works for most of the cases, but not for monthly calendar, for instance

```py
time_index = pd.date_range("20181001", "20211001", freq="2M")
df = pd.DataFrame(
    data={
        "data1": np.random.random(len(time_index)),
    },
    index=time_index,
)

cal = monthly_calendar(anchor="10-15", length="1M")
cal = cal.map_to_data(df)
lilio.resample(cal, df)
```
This will throw the following warning messages, which indeed have nothing to do with frequency, but only NaN values.
```py
/home/yangliu/AI4S2S/lilio/lilio/utils.py:104: FutureWarning: Units 'M', 'Y' and 'y' do not represent unambiguous timedelta values and will be removed in a future version.
  return pd.Timedelta(data_freq)
/home/yangliu/miniconda3/envs/s2spy/lib/python3.10/site-packages/numpy/core/fromnumeric.py:3474: RuntimeWarning: Mean of empty slice.
  return _methods._mean(a, axis=axis, dtype=dtype,
/home/yangliu/miniconda3/envs/s2spy/lib/python3.10/site-packages/numpy/core/_methods.py:189: RuntimeWarning: invalid value encountered in double_scalars
  ret = ret.dtype.type(ret / rcount)
/home/yangliu/AI4S2S/lilio/lilio/utils.py:80: UserWarning: The input data could not fully cover the calendar's intervals. Intervals without available data will contain NaN values.
  warnings.warn(
```

As we have provided `monthly_calendar` in the shorthands, it would be nice to have check for it as well.

File: lilio/utils.py [hunk 1]
```diff
@@ -122,10 +164,48 @@ def convert_interval_to_bounds(data: xr.Dataset) -> xr.Dataset:
...
+    data["left_bound"] = ("coord", bounds[:, 0])
+    data["right_bound"] = ("coord", bounds[:, 1])
+    data["left_bound"].attrs = {
+        "name": "Left bound of the interval",
+        "closed": "True",
+    }
+    data["right_bound"].attrs = {
+        "name": "Right bound of the interval",
+        "closed": "False",
+    }
```
[the remark this answers is not in the document]
Comment (BSchilperoort) [reply lilio/utils.py:174]: It should be visible in the HTML representation, as well as in a NetCDF file. We probably should add more attr info to the calendar resample output, such as storing the calendar used to generate it.

File: lilio/utils.py [hunk 2]
```diff
@@ -22,11 +27,11 @@ def check_timeseries(
      - Input data has a time index (pd), or a dim named `time` containing datetime
        values
     """
-    if not isinstance(data, PandasData + XArrayData):
+    if not isinstance(data, (pd.Series, pd.DataFrame, xr.DataArray, xr.Dataset)):
```
[the remark this answers is not in the document]
Comment (BSchilperoort) [reply lilio/utils.py:26]: I've decided to stop using the aliases, it does not save that much code, can make it less clear for what is happening, and I don't like them that much in the auto api docs.

I now removed PandasData and XArrayData everywhere in lilio.

**Comment by BSchilperoort** [comment]
> The checks for frequency works for most of the cases, but not for monthly calendar, for instance

Thanks for spotting this bug, Yang. It turns out that the Calendar's frequency is checked correctly, but the data's frequency is not! I'll modify the check to account for monthly-freq data inputs.

**Pull request merged**